src/main/kotlin/com/fasterxml/jackson/module/kotlin/Extensions.kt

```diff
@@ -50,21 +51,58 @@ fun ObjectMapper.registerKotlinModule(initializer: KotlinModule.Builder.() -> Un
 
 inline fun <reified T> jacksonTypeRef(): TypeReference<T> = object: TypeReference<T>() {}
 
+/**
+ * It is public due to Kotlin restrictions, but should not be used externally.
+ */
+inline fun <reified T> Any?.checkTypeMismatch(): T {
+    // Basically, this check assumes that T is non-null and the value is null.
+    // Since this can be caused by both input or ObjectMapper implementation errors,
+    // a more abstract RuntimeJsonMappingException is thrown.
+    if (this !is T) {
+        // Since the databind implementation of MappingIterator throws RuntimeJsonMappingException,
+        // JsonMappingException was not used to unify the behavior.
+        throw RuntimeJsonMappingException(
```

Comment:
As to what you discussed with #934, I personally felt that `JsonMappingException` would still be appropriate.
However, as commented, there was a problem with `MappingIterator` wrapping `JsonMappingException` with `RuntimeJsonMappingException`, so I finally chose `RealtimeJsonMappingException`.
https://github.com/FasterXML/jackson-databind/blob/25e1135e96cd78b775fb6de70beacad3e87d4800/src/main/java/com/fasterxml/jackson/databind/MappingIterator.java#L425-L427

How do you like this decision?

Reply:
When considering `RuntimeException` vs `CheckedException`. 
`RuntimeException` 👍🏼

Reply:
One thing to notice when merging to master (3.0): all `JacksonException`s are unchecked.
So can use more appropriate exception type for 3.0.

As to 2.x, wherever checked base type (`IOException`) declared, I think that is better choice than `RuntimeException`, since caller will need to be prepared to catch/handle that anyway.

Reply:
Thanks for your comment, for `2.19` I will use `RuntimeJsonMappingException`.

> One thing to notice when merging to master (3.0): all JacksonExceptions are unchecked.
So can use more appropriate exception type for 3.0.

I checked and it is not wrapped with `RealtimeJsonMappingException` in `3.x`.
I will consider what is appropriate to use in `3.x`.